### PR TITLE
Support nopython sets

### PR DESCRIPTION
**Pull request by pitrou**
Support for sets in nopython mode. This requires hash() to be defined on the key type: currently on numbers, booleans and tuples.

All set operations are supported. However, sets cannot (as of this PR) be passed from or to Python code.

### CODE REVIEW

File: docs/source/reference/pysupported.rst [hunk 1]
```diff
@@ -112,6 +112,16 @@ contains a :class:`int` and a :class:`float`).
...
+set
+---
+
+All methods and operations on sets are supported in JIT-compiled functions.
+However, passing and returning sets from/to Python code is not supported
+(yet).
+
+Sets must be strictly homogenous: Numba will reject any set containing
+objects of different types, even if the types are compatible (for example,
+``{1, 2.5}`` is rejected as it contains a :class:`int` and a :class:`float`).
```
Comment (stuartarchibald) [inline docs/source/reference/pysupported.rst]: ```:class`int`

Comment (pitrou) [inline docs/source/reference/pysupported.rst]: Hmm, how so?

Comment (stuartarchibald) [inline docs/source/reference/pysupported.rst]: Might just be the github renderer https://github.com/pitrou/numba/blob/setobj/docs/source/reference/pysupported.rst

Comment (pitrou) [inline docs/source/reference/pysupported.rst]: Oh, right, Sphinx should link to Python's `int` documentation.

File: numba/cgutils.py [hunk 2]
```diff
@@ -932,18 +937,16 @@ def memcpy(builder, dst, src, count):
...
         in_ptr = builder.gep(src, [loop.index])
         builder.store(builder.load(in_ptr), out_ptr)
 
 
-def memmove(builder, dst, src, count, itemsize, align=1):
-    """
-    Emit a memmove() call for `count` items of size `itemsize`
-    from `src` to `dest`.
-    """
+def _raw_memcpy(builder, func_name, dst, src, count, itemsize, align):
```
Comment (stuartarchibald) [inline numba/cgutils.py]: does the body of this function need updating to use `func_name`? https://github.com/pitrou/numba/blob/setobj/numba/cgutils.py#L953 suggests `llvm.memmove` is still in use?

Comment (pitrou) [inline numba/cgutils.py]: Oh, right... Thanks for noticing.

File: numba/runtime/nrt.c [hunk 3]
```diff
@@ -356,6 +356,22 @@ NRT_MemInfo *NRT_MemInfo_varsize_alloc(size_t size)
     return mi;
 }
 
+void *NRT_MemInfo_varsize_alloc(NRT_MemInfo *mi, size_t size)
+{
+    if (mi->dtor != nrt_varsize_dtor) {
```
Comment (stuartarchibald) [inline numba/runtime/nrt.c]: Can `*mi` be `NULL` or size 0 ?

Comment (pitrou) [inline numba/runtime/nrt.c]: No :-)

File: numba/tests/test_builtins.py [hunk 4]
```diff
@@ -77,6 +77,8 @@ def format_usecase(x, y):
 def globals_usecase():
     return globals()
 
+# NOTE: hash() is tested is test_hashing
```
Comment (stuartarchibald) [inline numba/tests/test_builtins.py]: `tested in test_hashing.py`?

Comment (pitrou) [inline numba/tests/test_builtins.py]: Oh, you're right.

File: numba/tests/test_sets.py [hunk 5]
```diff
@@ -1,42 +1,516 @@
...
+
+    def test_clear(self):
+        pyfunc = clear_usecase
+        check = self.unordered_checker(pyfunc)
+
+        check((1, 2, 4, 11))
+        check(self.sparse_array(50))
+
+    def test_copy(self):
+        # Source set doesn't have any deleted entries
```
Comment (stuartarchibald) [inline numba/tests/test_sets.py]: Worth a test of copying an empty set?

Comment (pitrou) [inline numba/tests/test_sets.py]: Yes, probably.

**Comment by seibert** [comment]
Looks like all the comments have been addressed.

**Pull request merged**